File: digital-banking-web/src/app/services/user-service.ts
import { Injectable } from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {environment} from '../../environments/environment';
import {ChangePasswordDTO, ResetPasswordAdminDTO, UserRequestDTO, UserResponseDTO} from '../model/user.model';
import {Observable} from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class UserService {

  constructor(private http: HttpClient) { }

  createUser(user: any) {
    return this.http.post(`${environment.backendHost}/api/users/createUser`, user);
  }

  // GET BY ID
  getUserById(id: number): Observable<UserResponseDTO> {
    return this.http.get<UserResponseDTO>(`${environment.backendHost}/api/users/${id}`);
  }

  // GET BY USERNAME
  getUserByUsername(username: string): Observable<UserResponseDTO> {
    return this.http.get<UserResponseDTO>(`${environment.backendHost}/api/users/username/${username}`);
  }

  // GET ALL USERS
  getAllUsers(): Observable<UserResponseDTO[]> {
    return this.http.get<UserResponseDTO[]>(`${environment.backendHost}/api/users/all`);
  }

  // UPDATE
  updateUser(id: number, user: UserRequestDTO): Observable<UserResponseDTO> {
    return this.http.put<UserResponseDTO>(`${environment.backendHost}/api/users/${id}`, user);
  }

  // DELETE
  deleteUser(id: number): Observable<void> {
    return this.http.delete<void>(`${environment.backendHost}/api/users/${id}`);
  }

  // resetPasswordAdmin
  resetPasswordAdmin(id: number, resetPasswordAdminDTO: ResetPasswordAdminDTO) {
    return this.http.patch(
      `${environment.backendHost}/api/users/admin/${id}/reset-password`,
        resetPasswordAdminDTO,
      { responseType: 'text' }
    );
  }

  changePassword(data : ChangePasswordDTO) {
    return this.http.patch(
      `${environment.backendHost}/api/users/change-password`,
      data,
      {
        responseType: 'text',
      },
    );
  }


}

File: ebanking-backend/src/main/java/ma/saifdine/hd/ebankingbackend/services/UserService.java
package ma.saifdine.hd.ebankingbackend.services;

import ma.saifdine.hd.ebankingbackend.dtos.ChangePasswordRequestDTO;
import ma.saifdine.hd.ebankingbackend.dtos.ResetPasswordAdminDTO;
import ma.saifdine.hd.ebankingbackend.dtos.UserRequestDTO;
import ma.saifdine.hd.ebankingbackend.dtos.UserResponseDTO;

import java.util.List;

public interface UserService {

    UserResponseDTO createUser(UserRequestDTO userRequestDTO);

    UserResponseDTO getUserByUsername(String username);

    UserResponseDTO getUserById(long id);

    UserResponseDTO updateUser(Long id,UserRequestDTO userRequestDTO);

    void deleteUser(Long id);

    List<UserResponseDTO> getAllUsers();

    void changePassword(String username, ChangePasswordRequestDTO changePasswordRequestDTO);

    void adminResetPassword(Long id, ResetPasswordAdminDTO passwordAdminDTO);

}
